crypto/evp/pmeth_lib.c

```diff
@@ -878,6 +878,12 @@ static int legacy_ctrl_to_param(EVP_PKEY_CTX *ctx, int keytype, int optype,
             return EVP_PKEY_CTX_set_rsa_keygen_pubexp(ctx, p2);
         case EVP_PKEY_CTRL_RSA_KEYGEN_PRIMES:
             return EVP_PKEY_CTX_set_rsa_keygen_primes(ctx, p1);
+        case EVP_PKEY_CTRL_DSA_PARAMGEN_BITS:
+            return EVP_PKEY_CTX_set_dsa_paramgen_bits(ctx, p1);
+        case EVP_PKEY_CTRL_DSA_PARAMGEN_Q_BITS:
+            return EVP_PKEY_CTX_set_dsa_paramgen_q_bits(ctx, p1);
+        case EVP_PKEY_CTRL_DSA_PARAMGEN_MD:
+            return EVP_PKEY_CTX_set_dsa_paramgen_md(ctx, p2);
```

Comment:
It seems there's an overlap between EVP_PKEY_CTRL_DSA and EVP_PKEY_CTRL_RSA values.  `keytype == -1` is reaaaallllllly dodgy in this regard, we should probably work to avoid that if we can

Reply:
I have just remapped the DSA ones to start at + 20. is that ok?

Reply:
Grudgingly...  we really shouldn't place these in the keytype == -1 section

Reply:
Oh, I now saw that you're (*correctly*) calling `EVP_PKEY_CTX_ctrl()` with `EVP_PKEY_DSA`, which is not -1, so these cases are wrongly placed.  How they are even reached, I currently cannot fathom...

...
and I realise I've done the same stoopid misplacement, so I totally get that you merely followed bad example.

So, uhm, no.  Not ok.

Reply:
#11329 fixes my mistakes.

Reply:
Updated